### ISSUE TITLE
WasmFS: IndexedDB backend

### ISSUE THREAD

**Issue by tlively** [emscripten-core/emscripten#15942]
Add a concrete WasmFS backend for managing files in IndexedDB, analogous to the current [IDBFS](https://emscripten.org/docs/api_reference/Filesystem-API.html#filesystem-api-idbfs).

**Comment by junyuecao** [comment]
The current [IDBFS](https://emscripten.org/docs/api_reference/Filesystem-API.html#filesystem-api-idbfs) only provide file persistence and it loads all file into memory at once. It'll be great if the new FS could load into memory by demand.

**Comment by tlively** [comment]
Yes, the way I envision this working is that the IDB backend itself would always load data on-demand and a separately configured caching layer would improve performance by keeping a limited amount of data in memory.

**Comment by junyuecao** [comment]
@tlively May I ask for the current status of IDB backend or what is the plan?

**Comment by tlively** [comment]
I've written a design doc for how a WasmFS IndexedDB backend might work, but no implementation work has started on it. So far we've been prioritizing work on the OPFS backend to get it to be stable and fast. Even with a better IndexedDB backend design, OPFS with SyncAccessHandles will still be much faster than IndexedDB. Do you have a use case for IndexedDB that couldn't use OPFS?

**Comment by mere-human** [comment]
I have a use case for IndexedDB backend. I'm using it for saving some user modifications between browser sessions in a document editing app. But currently it's done via custom FS backend + Emscripten async IndexedDB API (`emscripten_idb_async_delete()` and such).

**Comment by maccman** [comment]
I think it's important that we have a fallback to SyncAccessHandles when they are not available. I believe IndexedDb is the only viable one because it doesn't have storage limitations - but I could be wrong?

**Comment by AntonOfTheWoods** [comment]
There is also the fact that `sync` also only works in web workers, and there are other places, like extensions, where it would be great to have persistent storage!

**Comment by defundinformants** [comment]
wait you think this project is open? it was ended a long tine ago. stop and abort everything before you get introuble

**Comment by tlively** [comment]
The latest thinking here is that it would be good to provide a bridge between WasmFS and the legacy file system backends, including IDBFS, rather than implementing a new WasmFS IndexedDB backend from scratch (although that would certainly still be possible and would have better performance). Unfortunately no concrete work is planned for this, though.

**Comment by mere-human** [comment]
> The latest thinking here is that it would be good to provide a bridge between WasmFS and the legacy file system backends

@tlively Is there a GitHub issue where this work could be tracked?

**Comment by tlively** [comment]
@mere-human the issue for this is https://github.com/emscripten-core/emscripten/issues/18935.
